### PR TITLE
Clarify optionality support for the double precision literals.

### PR DESCRIPTION
**Pull request by AnastasiaStulova**
This change adds clarifications regarding double precision
literals when double precision is not supported.

It also improves wording for -cl-single-precision-constant
flag to make it more precise.

### CODE REVIEW

**Comment by AnastasiaStulova** [comment]
@b-sumner and @jlewis-austin I was not able to request your review... Perhaps you are not in OpenCL group yet?

**Comment by jlewis-austin** [comment]
> @b-sumner and @jlewis-austin I was not able to request your review... Perhaps you are not in OpenCL group yet?

@AnastasiaStulova thanks I'll take a look, and emailed James regarding getting us added.

File: OpenCL_C.txt [hunk 1]
```diff
@@ -307,7 +307,8 @@ The following table describes the list of built-in scalar data types.
       The `double` data type must conform to the IEEE 754 double precision
       storage format.
 
-      <<unified-spec, Requires>> support for OpenCL C 1.2 or newer.
+      <<unified-spec, Requires>> support for OpenCL C 1.2 or newer. In
+      OpenCL C 3.0 it requires support of {opencl_c_fp64} feature.
```
Comment (jlewis-austin) [inline OpenCL_C.txt:311]:
```suggestion
      OpenCL C 3.0 it requires support of the {opencl_c_fp64} feature.
```
Elsewhere it looks like we refer to the feature macro rather than the feature. I'm not sure if there is any distinction in meaning?

File: OpenCL_C.txt [hunk 2]
```diff
@@ -307,7 +307,8 @@ The following table describes the list of built-in scalar data types.
       The `double` data type must conform to the IEEE 754 double precision
       storage format.
 
-      <<unified-spec, Requires>> support for OpenCL C 1.2 or newer.
+      <<unified-spec, Requires>> support for OpenCL C 1.2 or newer. In
```
Comment (jlewis-austin) [inline OpenCL_C.txt:310]: Should this change be duplicated for the double vector types?

**Review by jlewis-austin — COMMENTED** [review]
Thanks, these look good, just a few editorial suggestions.

File: OpenCL_C.txt (hunk 1 above)
Comment (AnastasiaStulova) [reply OpenCL_C.txt:311]: Ah I think it was done when we didn't have wording saying that features inherit names from the macros but now we do. For extensions we used to refer to the name rather than the macro. This is why we wanted to update the features but it has not been done yet.

I can reword it to alight with the other features and we change all of them later or I can keep it as is since this is what we aim at but it will be inconsistent with the other features but consistent with extensions...

File: OpenCL_C.txt (hunk 2 above)
Comment (AnastasiaStulova) [reply OpenCL_C.txt:310]: Done! Thanks!

**Comment by AnastasiaStulova** [comment]
FYI I also fixed spelling of `double precision` and `single precision` in various places since it was inconsistent. I can undo it though.

File: OpenCL_C.txt (hunk 1 above)
Comment (jlewis-austin) [reply OpenCL_C.txt:311]: OK thanks, however you want to stage it is fine with me

**Comment by jlewis-austin** [comment]
> FYI I also fixed spelling of `double precision` and `single precision` in various places since it was inconsistent. I can undo it though.

The rule I've seen most often elsewhere is to hyphenate when used as an adjective, e.g. "single-precision floating-point literal" vs "operator supports single precision floating point". That being said, I'm OK with any choice consistently applied.

File: OpenCL_C.txt (hunk 1 above)
Comment (bashbaug) [reply OpenCL_C.txt:311]: I do think we should accept this suggestion though and add `the`.

**Comment by bashbaug** [comment]
> Maybe we can use the most common spelling in this PR, but it can be fully renamed separately?

Yes, absolutely, we shouldn't "fix" the hyphenation issues in this PR, but I'm hoping we don't introduce new issues either.

We can add these terms to this issue if you'd like: https://github.com/KhronosGroup/OpenCL-Docs/issues/474

Remember: we can't simply use the hyphenated or non-hyphenated version exclusively since the correct version depends on how it is used.

File: OpenCL_C.txt [hunk 3]
```diff
@@ -332,6 +333,12 @@ The following table describes the list of built-in scalar data types.
       type that cannot be completed.
 |====
 
+If double precision floating-point extension {cl_khr_fp64} or feature
+{opencl_c_fp64} is not supported, implementations are allowed to
+implicitly cast double precision floating-point literals to
+single precision literals. The use of double precision literals without
```
Comment (AnastasiaStulova) [inline OpenCL_C.txt:339]:
```suggestion
single-precision literals. The use of double-precision literals without
```

File: OpenCL_C.txt [hunk 4]
```diff
@@ -332,6 +333,12 @@ The following table describes the list of built-in scalar data types.
...
+double precision support should result in a diagnostic.
```
Comment (AnastasiaStulova) [inline OpenCL_C.txt:340]:
```suggestion
double-precision support should result in a diagnostic.
```

File: OpenCL_C.txt [hunk 5]
```diff
@@ -332,6 +333,12 @@ The following table describes the list of built-in scalar data types.
       type that cannot be completed.
 |====
 
+If double precision floating-point extension {cl_khr_fp64} or feature
+{opencl_c_fp64} is not supported, implementations are allowed to
+implicitly cast double precision floating-point literals to
```
Comment (AnastasiaStulova) [inline OpenCL_C.txt:338]:
```suggestion
implicitly cast double-precision floating-point literals to
```

File: OpenCL_C.txt [hunk 6]
```diff
@@ -474,7 +481,8 @@ The following table describes the list of built-in vector data types.
 | `double__n__` footnote:[{fn-double-vec}]
     | A vector of _n_ 64-bit floating-point values.
 
-      <<unified-spec, Requires>> support for OpenCL C 1.2 or newer.
+      <<unified-spec, Requires>> support for OpenCL C 1.2 or newer. In
+      OpenCL C 3.0 it requires support of {opencl_c_fp64} feature.
```
Comment (AnastasiaStulova) [inline OpenCL_C.txt:485]:
```suggestion
      OpenCL C 3.0 it requires support of the {opencl_c_fp64} feature.
```

File: OpenCL_C.txt [hunk 7]
```diff
@@ -4814,7 +4822,7 @@ all arguments and the return type, unless otherwise specified.
     | Returns a quiet NaN.
       The _nancode_ may be placed in the significand of the resulting NaN.
 | gentype *nextafter*(gentype _x_, gentype _y_)
-    | Computes the next representable single-precision floating-point value
+    | Computes the next representable single precision floating-point value
```
[the remark this answers is not in the document]
Comment (AnastasiaStulova) [reply OpenCL_C.txt:4825]:
```suggestion
    | Computes the next representable single-precision floating-point value
```

File: OpenCL_C.txt [hunk 8]
```diff
@@ -5056,7 +5064,7 @@ single precision floating-point number.
 |====
 | *Constant Name* | *Description*
 | `MAXFLOAT`
-    | Value of maximum non-infinite single-precision floating-point number.
+    | Value of maximum non-infinite single precision floating-point number.
```
Comment (AnastasiaStulova) [inline OpenCL_C.txt:5067]:
```suggestion
    | Value of maximum non-infinite single-precision floating-point number.
```

File: OpenCL_C.txt [hunk 9]
```diff
@@ -8181,7 +8189,7 @@ a following *a*, *A*, *e*, *E*, *f*, *F*, *g*, or *G* conversion specifier
 applies to a `double__n__` argument.
 The *l* modifier is supported by the full profile.
 For the embedded profile, the *l* modifier is supported only if 64-bit
-integers or double-precision floating-point are supported by the device.
+integers or double precision floating-point are supported by the device.
```
Comment (AnastasiaStulova) [inline OpenCL_C.txt:8192]:
```suggestion
integers or double-precision floating-point are supported by the device.
```

File: api/opencl_runtime_layer.asciidoc [hunk 10]
```diff
@@ -6211,8 +6211,8 @@ These options control compiler behavior regarding floating-point arithmetic.
 These options trade off between speed and correctness.
 
 `-cl-single-precision-constant` ::
-    Treat double precision floating-point constant as single precision
-    constant.
+    This option forces implicit conversions of double precision floating-point
```
Comment (AnastasiaStulova) [inline api/opencl_runtime_layer.asciidoc:6214]:
```suggestion
    This option forces implicit conversions of double-precision floating-point
```

**Comment by AnastasiaStulova** [comment]
> Remember: we can't simply use the hyphenated or non-hyphenated version exclusively since the correct version depends on how it is used.

Thanks for the pointer. It really helped to remove the confusion. I think I have fixed those issues now.

**Review by bashbaug — APPROVED** [review]
Thanks LGTM!
